sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp

```diff
@@ -102,16 +103,39 @@ cdef DTYPE_t[::1] _sqeuclidean_row_norms32_dense(
     return squared_row_norms
 
 
+cdef DTYPE_t[::1] _sqeuclidean_row_norms64_sparse(
+    const DTYPE_t[:] X_data,
+    const SPARSE_INDEX_TYPE_t[:] X_indptr,
+    ITYPE_t num_threads,
+):
+    cdef:
+        ITYPE_t n = X_indptr.shape[0] - 1
+        ITYPE_t idx = 0
+        ITYPE_t X_i_ptr
+        DTYPE_t[::1] squared_row_norms = np.zeros(n, dtype=DTYPE)
+
+    for idx in prange(n, schedule='static', nogil=True, num_threads=num_threads):
+        for X_i_ptr in range(X_indptr[idx], X_indptr[idx+1]):
+            squared_row_norms[idx] += X_data[X_i_ptr] * X_data[X_i_ptr]
+
+    return squared_row_norms
+
+
 {{for name_suffix, INPUT_DTYPE_t, INPUT_DTYPE in implementation_specific_values}}
 
 from ._datasets_pair cimport DatasetsPair{{name_suffix}}
 
 
 cpdef DTYPE_t[::1] _sqeuclidean_row_norms{{name_suffix}}(
-    const {{INPUT_DTYPE_t}}[:, ::1] X,
+    X,
     ITYPE_t num_threads,
 ):
-    return _sqeuclidean_row_norms{{name_suffix}}_dense(X, num_threads)
+    if issparse(X):
+        X_data = np.asarray(X.data, dtype=DTYPE)
```

Comment:
`X.data` is be cast twice to float64 if it's float32 (here and in the constructor). Is there a way to only cast it once?

Same comment regarding `Y.data`.

Reply:
I don't see any obvious way to only cast once since the prior casting occurs within `MiddleTermComputer`. As you suggested, we could bring `sqeuclidean_row_norm` under the `MiddleTermComputer` umbrella in a future PR. This way, we would benefit from a unified logic.

Reply:
OK, this is definitely a good motivation for this refactoring.

I would also have this refactoring done as part of another PR. If this PR is merge, relying on the CSR-CSR case for `metric="euclidean"` in float32 would come with an extra non-negligible yet not too costly duplicated allocation and cast.

I would wait for another review's opinion before merging this PR.

Reply:
```suggestion
        # TODO: remove this instruction which is a cast in the float32 case
        # by moving squared row norms computations in MiddleTermComputer. 
        X_data = np.asarray(X.data, dtype=DTYPE)
```